Adds ENS support

wackerow opened the pull request:
🚨 `wallet-projects` (PR #307) <- `profile-layout`
https://www.notion.so/efdn/Display-user-s-ENS-address-once-connected-9aeaf51a811a4573be25a1a2852cd4f1

### Description
- Adds ENS related functions to `/utils/accounts` (utilizes `VUE_APP_MAINNET_PROVIDER_URL` .env variable)
- Looks up possible ENS for current user address upon loading/setting user (`LOAD_USER_INFO`)
- `ensName` property added to `User` interface to store this data
- Updated Profile and WalletWidget to show ENS if available. In Profile, if no ENS available, redundant 0x address is hidden. 
- Layout adjustments for Profile on mobile
- Restyles addresses to use `text-overflow: ellipsis;` to improve responsiveness and flexibility

**Join** form changes:
- Adds ENS support to team funding address input.
- Utilizes `v-model.lazy` to only validate this field upon leaving/blurring to prevent ENS queries on every keystroke.
- Applied this same `.lazy` approach to existing issues with "email" input (only validates now after user leaves). 
- Form summary shows ENS if available, and in that case also shows the corresponding address dimmed below for confirmation.
- ProjectProfile used for displaying project and project preview shows the ENS address if available, but link and copy button reference the resolved 0x address.

### Screenshots
![image](https://user-images.githubusercontent.com/54227730/131199802-f5c01555-6a73-4831-9a8d-a608a52e2773.png)
![image](https://user-images.githubusercontent.com/54227730/131199838-853533be-e7c3-4a57-a455-e8670e6d42d6.png)

wackerow committed:
set width to 100% on mobile

wackerow committed:
adjust target width

wackerow committed:
Adds ENS lookup for Profile

ENS requires tapping into .env mainnet provider.
Switches displayed address to end-of-string ellipsis for overflow. More dynamic, flexible and simplified this way.
Hides redundant 0x address if no ENS available.

wackerow committed:
Adds ENS support to Join form

wackerow committed:
implements v-model.lazy to Join form

Delays validation of URL, email and address forms until leaving field

wackerow committed:
adds account validator

also cleans up existing logic and adds some comments

wackerow committed:
logic to resolve possible ENS - join form

validates address field for either 0x address or ENS, stores `hasEns` boolean if user entered a valid ENS name, and adjusts how this is displayed on summary page depending on circumstance.

wackerow committed:
logic to display ENS if available on project profile

wackerow committed:
Merge branch 'wallet-projects' into profile-layout

wackerow committed:
Merge branch 'join-ens' into profile-layout

wackerow committed:
call ensLookup from store on user load, not from component

Adds `ensName` property to `User` interface. Looks up ENS when setting the user in the store. Prevents profile from re-fetching data every time it is opened.

wackerow committed:
show ENS in WalletWidget

Caps width to smaller of either 14 character, or 20 vw (for mobile). Simplifies overflow to trailing ellipsis.

samajammin commented on vue-app/src/api/core.ts:
```diff
@@ -2,6 +2,9 @@ import { ethers } from 'ethers'
 import { FundingRoundFactory } from './abi'
 
+export const mainnetProvider = new ethers.providers.StaticJsonRpcProvider(
+  process.env.VUE_APP_MAINNET_PROVIDER_URL
```
```suggestion
  process.env. VUE_APP_ETHEREUM_MAINNET_API_URL
```
I suggest sticking with the "API" terminology for consistency.

samajammin replied:
Also please add this ENV variable to the .env.example file.

samajammin commented on vue-app/src/views/Profile.vue:
```diff
@@ -130,10 +132,14 @@ export default class NavBar extends Vue {
     this.isLoading = true
     await this.loadProjects()
+    // if (this.currentUser) {
+    //   this.ens = await ensLookup(this.currentUser.walletAddress)
```
This probably shouldn't be commented out 😄

wackerow replied:
you're right, it should be gone 😆... this is stored in the state now
(see `get displayAddress()`)

samajammin commented on vue-app/src/store/index.ts:
```diff
@@ -327,11 +328,17 @@ const actions = {
       }
 
+      let ensName: string | null = state.currentUser.ensName
+      if (isAddress(state.currentUser.walletAddress)) {
```
Can we safely assume this is an address at this point? If not, we probably should add this check sooner (e.g. before querying ether & token balances).

samajammin replied:
Perhaps @pettinarip can confirm but I think we can safely assume this is a legit address. I think we only set `state.currentUser.walletAddress` from the Web3 plugin, which receives this from the connected accounts.

wackerow replied:
Yeah, I think you're right... This is pulled from the wallet provider, will remove check

samajammin commented on vue-app/src/utils/accounts.ts:
```diff
@@ -1,5 +1,32 @@
+  return await mainnetProvider.resolveName(name)
+}
+
+export async function isValidEns(name: string): Promise<boolean> {
```
Doesn't look like we use this anywhere - should we be?

wackerow replied:
Good catch... I was only using it in an existing approach to the form validation. This is no longer needed with `isValidEthAddress` as a replacement.

samajammin commented on vue-app/src/utils/accounts.ts:
```diff
@@ -1,5 +1,32 @@
+}
+
+export async function isValidEthAddress(address: string): Promise<boolean> {
+  // Returns true if address is valid ENS or 0x address
```
Love to see the comments! Minor nitpick - generally comments describing functions should go on the line above the function declaration.

samajammin commented on vue-app/src/views/Profile.vue:
```diff
@@ -153,22 +159,9 @@ export default class NavBar extends Vue {
     return CHAIN_INFO[Number(process.env.VUE_APP_ETHEREUM_API_CHAINID)]
   }
 
-  renderUserAddress(digitsToShow?: number): string {
```
Don't we want to use this if we're displaying an address? I think this change created this issue: https://github.com/ethereum/clrfund/issues/327

wackerow replied:
@samajammin So I tried switching this out for just using CSS to end with an ellipsis, allowing some more flexibility with what gets placed here, and it's just simpler to deal with that this helper function. 

In issue #327, this is all due to not rounding the decimals to 5 in this component, and can be addressed separately. Don't this directly caused that issue. 

If there is a stylistic preference to not let the 0x addresses get too long, we can still place a character limit on this element to start the ellipsis sooner. Let me know what you think, otherwise I'll leave as it.

samajammin replied:
Ah good call, didn't see the CSS change 👍

samajammin reviewed (changes requested):
Nice work! A few things I believe we should change - see comments.

wackerow committed:
update vue-app/.env.example

changes mainnet RPC variable name to `VUE_APP_ETHEREUM_MAINNET_API_URL`

wackerow committed:
remove local ens state

in lieu of $store.state which holds ens info

wackerow committed:
removes check for `isAddress`

Can safely assume as this is fetched from wallet provider itself

wackerow committed:
removes unused method

No longer needed now that `isValidEthAddress` is being used in its place

wackerow committed:
move comments above function declaration

wackerow committed:
adds `requester` property to `Project` interface

Allows showing projects in Profile that a user may have submitted as the requester, even if they are not the beneficiary address.

samajammin reviewed (approved):
Awesome work on this 💪
